File: 6_kyu/Reducing_by_steps.jl
module Step
    export oper_array
    export gcdi
    export lcmu
    export som
    export maxi
    export mini

    gcdi = gcd
    lcmu = lcm
    som = +
    maxi = max
    mini = min
    
    oper_array(fct, arr, init) = map( x -> init=fct(init,x), arr )

end

File: 6_kyu/Reducing_by_steps.rs
use num::integer;

fn som(x: i64, y: i64) -> i64 {
    x + y
}
fn maxi(x: i64, y: i64) -> i64 {
    x.max(y)
}
fn mini(x: i64, y: i64) -> i64 {
    x.min(y)
}
fn gcdi(m: i64, n: i64) -> i64 {
    integer::gcd(m,n)
}
fn lcmu(a: i64, b: i64) -> i64 {
    integer::lcm(a,b)
}

fn oper_array( fct: fn(i64,i64) -> i64, a: &[i64], init: i64) -> Vec<i64> {
    let mut accu = init;
    a.iter().map(|&x| {accu = fct(x,accu); accu} ).collect()
}
